content-tracker/src/main.rs

```diff
@@ -0,0 +1,261 @@
+        if addr.derp_region().is_some() {
+            break;
+        }
+        if t0.elapsed() > Duration::from_secs(10) {
+            anyhow::bail!("timeout waiting for DERP region");
+        }
+        tokio::time::sleep(Duration::from_millis(50)).await;
+    }
+    Ok(())
+}
+
+async fn create_endpoint(
+    key: iroh_net::key::SecretKey,
+    port: u16,
+) -> anyhow::Result<MagicEndpoint> {
+    // let pkarr_relay_discovery = discovery::PkarrRelayDiscovery::new(key.clone(), PKARR_RELAY_URL.parse().unwrap());
+    let region_discover = discovery::HardcodedRegionDiscovery::new(2);
+    iroh_net::MagicEndpoint::builder()
+        .secret_key(key)
+        .discovery(Box::new(region_discover))
+        .alpns(vec![TRACKER_ALPN.to_vec()])
+        .bind(port)
+        .await
+}
+
+/// Accept an incoming connection and extract the client-provided [`NodeId`] and ALPN protocol.
+pub async fn accept_conn(
+    mut conn: quinn::Connecting,
+) -> anyhow::Result<(NodeId, String, quinn::Connection)> {
+    let alpn = get_alpn(&mut conn).await?;
+    tracing::info!("awaiting conn");
+    let conn = conn.await?;
+    tracing::info!("got conn");
+    let peer_id = get_peer_id(&conn)?;
+    Ok((peer_id, alpn, conn))
+}
+
+/// Write default options to a sample config file.
+fn write_defaults() -> anyhow::Result<()> {
+    let default_path = tracker_path(CONFIG_DEFAULTS_FILE)?;
+    crate::io::save_to_file(Options::default(), &default_path)?;
+    Ok(())
+}
+
+async fn server(args: ServerArgs) -> anyhow::Result<()> {
+    set_verbose(!args.quiet);
+    let rt = tokio::runtime::Handle::current();
+    let tpc = LocalPoolHandle::new(2);
+    let rt = iroh_bytes::util::runtime::Handle::new(rt, tpc);
+    let home = tracker_home()?;
+    log!("tracker starting using {}", home.display());
+    let key_path = tracker_path(SERVER_KEY_FILE)?;
+    let key = load_secret_key(key_path).await?;
+    let endpoint = create_endpoint(key, args.port).await?;
+    let config_path = tracker_path(CONFIG_FILE)?;
+    write_defaults()?;
+    let mut options = load_from_file::<Options>(&config_path)?;
+    options.make_paths_relative(&home);
+    let db = Tracker::new(options)?;
+    await_derp_region(&endpoint).await?;
+    let addr = endpoint.my_addr().await?;
+    log!("listening on {:?}", addr);
+    log!("tracker addr: {}\n", addr.node_id);
+    log!("usage:");
+    log!("tracker announce --tracker {} <tickets>", addr.node_id);
+    log!(
+        "tracker query --tracker {} <hash> or <ticket>",
+        addr.node_id
+    );
+    log!();
+    let db2 = db.clone();
+    let endpoint2 = endpoint.clone();
+    let _task = rt
+        .local_pool()
+        .spawn_pinned(move || db2.probe_loop(endpoint2));
+    while let Some(connecting) = endpoint.accept().await {
+        tracing::info!("got connecting");
+        let db = db.clone();
+        tokio::spawn(async move {
+            let Ok((pk, h, conn)) = accept_conn(connecting).await else {
```

Comment:
`s/h/alpn`, and `s/pk/node_id`?
I was confused what `h` is. Also, abort if the `alpn` is unexpected?
Edit: Right the latter is not needed because the endpoint is configured with a single ALPN. Maybe add a comment that you'd want to check the ALPN here if the endpoint supported multiple protocols. If this is to serve as an example for people to follow along I mean.